File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/appleskin/AppleSkinIntegration.java
package crabcraft.net.crabUtilities.appleskin;

import io.papermc.paper.event.world.WorldGameRuleChangeEvent;
import org.bukkit.GameRule;
import org.bukkit.GameRules;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerChangedWorldEvent;
import org.bukkit.event.player.PlayerRegisterChannelEvent;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.plugin.messaging.Messenger;
import org.jetbrains.annotations.NotNull;
import org.jspecify.annotations.Nullable;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Objects;

public final class AppleSkinIntegration implements Listener {

    static final String SATURATION_CHANNEL = "appleskin:saturation";
    static final String EXHAUSTION_CHANNEL = "appleskin:exhaustion";
    private static final String NATURAL_REGENERATION_CHANNEL =
            "appleskin:natural_regeneration";
    private static final List<String> CHANNELS = List.of(
            SATURATION_CHANNEL,
            EXHAUSTION_CHANNEL,
            NATURAL_REGENERATION_CHANNEL);

    private static @Nullable JavaPlugin plugin;
    private static @Nullable AppleSkinIntegration listener;
    private static volatile boolean enabled;
    private static volatile long generation;

    private AppleSkinIntegration() {
    }

    public static synchronized boolean enable(@NotNull JavaPlugin plugin) {
        if (!plugin.getConfig().getBoolean("mod-protocols.appleskin.enabled", true)) {
            disable(plugin);
            plugin.getLogger().info("AppleSkin integration disabled in config.");
            return false;
        }
        if (enabled) {
            return true;
        }

        AppleSkinIntegration.plugin = plugin;
        generation++;

        try {
            listener = new AppleSkinIntegration();
            plugin.getServer().getPluginManager().registerEvents(listener, plugin);

            final Messenger messenger = plugin.getServer().getMessenger();
            CHANNELS.forEach(channel ->
                    messenger.registerOutgoingPluginChannel(plugin, channel));

            enabled = true;
            plugin.getServer().getOnlinePlayers().forEach(player -> {
                if (player.getListeningPluginChannels().contains(SATURATION_CHANNEL)) {
                    new AppleSkinSyncTask(player);
                }
                if (player.getListeningPluginChannels().contains(NATURAL_REGENERATION_CHANNEL)) {
                    syncNaturalRegeneration(player);
                }
            });
            plugin.getLogger().info("AppleSkin integration enabled");
            return true;
        } catch (LinkageError | RuntimeException exception) {
            disable(plugin);
            plugin.getLogger().warning(
                    "AppleSkin integration could not initialise: " + exception.getMessage());
            return false;
        }
    }

    public static synchronized void disable(@NotNull JavaPlugin plugin) {
        invalidateTasks();

        if (listener != null) {
            HandlerList.unregisterAll(listener);
            listener = null;
        }
        final Messenger messenger = plugin.getServer().getMessenger();
        CHANNELS.forEach(channel ->
                messenger.unregisterOutgoingPluginChannel(plugin, channel));
    }

    public static boolean isEnabled() {
        return enabled;
    }

    static long currentGeneration() {
        return generation;
    }

    static boolean isCurrentGeneration(long expected) {
        return enabled && generation == expected;
    }

    static JavaPlugin plugin() {
        return Objects.requireNonNull(plugin, "AppleSkin integration is not enabled");
    }

    static void invalidateTasks() {
        enabled = false;
        generation++;
        plugin = null;
    }

    @EventHandler
    private void onPlayerRegisterChannel(final PlayerRegisterChannelEvent event) {
        if (event.getChannel().equals(SATURATION_CHANNEL)) {
            new AppleSkinSyncTask(event.getPlayer());
        } else if (event.getChannel().equals(NATURAL_REGENERATION_CHANNEL)) {
            syncNaturalRegeneration(event.getPlayer());
        }
    }

    @EventHandler(priority = EventPriority.MONITOR)
    private void onPlayerChangedWorld(final PlayerChangedWorldEvent event) {
        if (event.getPlayer().getListeningPluginChannels()
                .contains(NATURAL_REGENERATION_CHANNEL)) {
            syncNaturalRegeneration(event.getPlayer());
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    private void onGameRuleChange(final WorldGameRuleChangeEvent event) {
        if (event.getGameRule() != naturalRegenerationRule()) {
            return;
        }
        event.getWorld().getPlayers().forEach(player -> {
            if (player.getListeningPluginChannels().contains(NATURAL_REGENERATION_CHANNEL)) {
                sendNaturalRegeneration(
                        player,
                        Boolean.parseBoolean(event.getValue()));
            }
        });
    }

    private static void syncNaturalRegeneration(Player player) {
        sendNaturalRegeneration(
                player,
                Boolean.TRUE.equals(
                        player.getWorld().getGameRuleValue(naturalRegenerationRule())));
    }

    private static void sendNaturalRegeneration(Player player, boolean enabled) {
        player.sendPluginMessage(
                plugin(),
                NATURAL_REGENERATION_CHANNEL,
                ByteBuffer.allocate(1).put((byte) (enabled ? 1 : 0)).array());
    }

    private static GameRule<Boolean> naturalRegenerationRule() {
        return GameRules.NATURAL_HEALTH_REGENERATION;
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/appleskin/AppleSkinIntegration.kt
package crabcraft.net.crabUtilities.appleskin

import io.papermc.paper.event.world.WorldGameRuleChangeEvent
import org.bukkit.GameRules
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.event.player.PlayerRegisterChannelEvent
import org.bukkit.plugin.java.JavaPlugin

class AppleSkinIntegration private constructor() : Listener {
    @EventHandler
    private fun onPlayerRegisterChannel(event: PlayerRegisterChannelEvent) {
        when (event.channel) {
            SATURATION_CHANNEL -> AppleSkinSyncTask(event.player)
            NATURAL_REGENERATION_CHANNEL -> syncNaturalRegeneration(event.player)
        }
    }

    @EventHandler(priority = EventPriority.MONITOR)
    private fun onPlayerChangedWorld(event: PlayerChangedWorldEvent) {
        if (NATURAL_REGENERATION_CHANNEL in event.player.listeningPluginChannels) syncNaturalRegeneration(event.player)
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    private fun onGameRuleChange(event: WorldGameRuleChangeEvent) {
        if (event.gameRule != naturalRegenerationRule()) return
        event.world.players.forEach { player ->
            if (NATURAL_REGENERATION_CHANNEL in player.listeningPluginChannels) {
                sendNaturalRegeneration(player, event.value.toBoolean())
            }
        }
    }

    companion object {
        const val SATURATION_CHANNEL = "appleskin:saturation"
        const val EXHAUSTION_CHANNEL = "appleskin:exhaustion"
        private const val NATURAL_REGENERATION_CHANNEL = "appleskin:natural_regeneration"
        private val CHANNELS = listOf(SATURATION_CHANNEL, EXHAUSTION_CHANNEL, NATURAL_REGENERATION_CHANNEL)
        private var plugin: JavaPlugin? = null
        private var listener: AppleSkinIntegration? = null
        @Volatile private var enabled = false
        @Volatile private var generation = 0L

        @JvmStatic
        @Synchronized
        fun enable(plugin: JavaPlugin): Boolean {
            if (!plugin.config.getBoolean("mod-protocols.appleskin.enabled", true)) {
                disable(plugin)
                plugin.logger.info("AppleSkin integration disabled in config.")
                return false
            }
            if (enabled) return true
            this.plugin = plugin
            generation++
            try {
                val integration = AppleSkinIntegration()
                listener = integration
                plugin.server.pluginManager.registerEvents(integration, plugin)
                val messenger = plugin.server.messenger
                CHANNELS.forEach { messenger.registerOutgoingPluginChannel(plugin, it) }
                enabled = true
                plugin.server.onlinePlayers.forEach { player ->
                    if (SATURATION_CHANNEL in player.listeningPluginChannels) AppleSkinSyncTask(player)
                    if (NATURAL_REGENERATION_CHANNEL in player.listeningPluginChannels) syncNaturalRegeneration(player)
                }
                plugin.logger.info("AppleSkin integration enabled")
                return true
            } catch (exception: LinkageError) {
                return failedEnable(plugin, exception)
            } catch (exception: RuntimeException) {
                return failedEnable(plugin, exception)
            }
        }

        private fun failedEnable(plugin: JavaPlugin, exception: Throwable): Boolean {
            disable(plugin)
            plugin.logger.warning("AppleSkin integration could not initialise: ${exception.message}")
            return false
        }

        @JvmStatic
        @Synchronized
        fun disable(plugin: JavaPlugin) {
            invalidateTasks()
            listener?.let { HandlerList.unregisterAll(it) }
            listener = null
            val messenger = plugin.server.messenger
            CHANNELS.forEach { messenger.unregisterOutgoingPluginChannel(plugin, it) }
        }

        @JvmStatic fun isEnabled() = enabled

        @JvmStatic fun currentGeneration() = generation

        @JvmStatic fun isCurrentGeneration(expected: Long) = enabled && generation == expected

        @JvmStatic
        fun plugin(): JavaPlugin = java.util.Objects.requireNonNull(plugin, "AppleSkin integration is not enabled")!!

        @JvmStatic
        fun invalidateTasks() {
            enabled = false
            generation++
            plugin = null
        }

        private fun syncNaturalRegeneration(player: Player) {
            sendNaturalRegeneration(player, player.world.getGameRuleValue(naturalRegenerationRule()) == true)
        }

        private fun sendNaturalRegeneration(player: Player, enabled: Boolean) {
            player.sendPluginMessage(plugin(), NATURAL_REGENERATION_CHANNEL, byteArrayOf(if (enabled) 1 else 0))
        }

        private fun naturalRegenerationRule() = GameRules.NATURAL_HEALTH_REGENERATION
    }
}
